modules/core/js/src/main/scala/net/message/Scram.scala

```diff
@@ -0,0 +1,148 @@
+    val nonce = bufferToByteVector(crypto.randomBytes(32)).toBase64
+    clientFirstBareWithNonce(nonce)
+  }
+
+  def clientFirstBareWithNonce(nonce: String): ByteVector =
+    s"n=,r=${nonce}".bytesUtf8
+
+  case class ServerFirst(nonce: String, salt: ByteVector, iterations: Int)
+  object ServerFirst {
+    private val Pattern = """r=([\x21-\x2B\x2D-\x7E]+),s=([A-Za-z0-9+/]+={0,2}),i=(\d+)""".r
+
+    def decode(bytes: ByteVector): Option[ServerFirst] =
+      utf8.decodeValue(bytes.bits).toOption.flatMap {
+        case Pattern(r, s, i) =>
+          Some(ServerFirst(r, ByteVector.fromValidBase64(s), i.toInt))
+        case _ => 
+          None
+      }
+  }
+
+  case class ClientProof(value: String)
+
+  case class ClientFinalWithoutProof(channelBinding: String, nonce: String) {
+    override def toString: String = s"c=$channelBinding,r=$nonce"
+    def encode: ByteVector = toString.bytesUtf8
+    def encodeWithProof(proof: ClientProof): ByteVector = (toString ++ s",p=${proof.value}").bytesUtf8
+  }
+
+  case class Verifier(value: ByteVector)
+
+  case class ServerFinal(verifier: Verifier)
+  object ServerFinal {
+    private val Pattern = """v=([A-Za-z0-9+/]+={0,2})""".r
+    def decode(bytes: ByteVector): Option[ServerFinal] =
+      utf8.decodeValue(bytes.bits).toOption.flatMap {
+        case Pattern(v) =>
+          Some(ServerFinal(Verifier(ByteVector.fromValidBase64(v))))
+        case _ => 
+          None
+      }
+  }
+
+  private val crypto = js.Dynamic.global.require("crypto")
+
+  private def bufferToByteVector(buffer: js.Dynamic): ByteVector =
+    ByteVector.view(TypedArrayBuffer.wrap(
+      buffer.asInstanceOf[js.typedarray.ArrayBuffer],
+      buffer.byteOffset.asInstanceOf[Int],
+      buffer.byteLength.asInstanceOf[Int]
+    ))
+
+  private def byteVectorToUint8Array(vector: ByteVector): js.typedarray.Uint8Array = {
+    val bb = vector.toByteBuffer
+    val ab = if (bb.hasArrayBuffer())
+      bb.arrayBuffer()
+    else {
+      val ab = new js.typedarray.ArrayBuffer(bb.remaining())
+      TypedArrayBuffer.wrap(ab).put(bb)
+      ab
+    }
+    new js.typedarray.Uint8Array(ab)
+  }
+
+  private def HMAC(key: ByteVector, str: ByteVector): ByteVector = {
+    val mac = crypto.createHmac("sha256", byteVectorToUint8Array(key))
+    mac.update(byteVectorToUint8Array(str))
+    bufferToByteVector(mac.digest())
+  }
+
+  private def H(input: ByteVector): ByteVector = {
+    val hash = crypto.createHash("sha256")
+    hash.update(byteVectorToUint8Array(input))
+    bufferToByteVector(hash.digest())
+  }
+
+  private def Hi(str: String, salt: ByteVector, iterations: Int): ByteVector = {
+    // TODO It is unfortunate that we have to use a sync API here when an async is available
+    // To make the change here will require running an F[_]: Async up the hiearchy    
+    val salted = crypto.pbkdf2Sync(str, byteVectorToUint8Array(salt), iterations, 8 * 32, "sha256")
+    bufferToByteVector(salted).take(32)
```

Comment:
@mpilquist you wouldn't happen to have any guesses why this `take(32)` is necessary here, but not on JVM?